File: data/figures/chapter02/figure_2_5/plot_2_5.m
function plot_2_5(outputPath)
%PLOT_2_5 Reproduce Figure 2.5 from three CSV inputs beside this function.
%   plot_2_5(outputPath) exports PNG, PDF or SVG according to the extension.
% Chart contract: static retained FCC55 lattice geometry comparison.
% Question: how do the retained beam paths change when 31 nodes snap to skin?
% Six orthographic panels show original/final coordinates in oblique and side
% views, followed by equal-scale interface enlargements. Grain: 348 nodes,
% 399 actual CBEAM edges, 713 shell
% triangles. CSV-only inputs; no projection connectors or amplified motion.
% Hard two-root palette: orange beams/moved nodes, blue receiving facets;
% neutral shell edges. Palette matches the microscopic chapter figures.
% Open/filled markers distinguish original/final. Exported PNG inspected.
plotDirectory = fileparts(mfilename('fullpath'));
if nargin < 1 || isempty(outputPath)
    outputPath = fullfile(plotDirectory,'figure_2_5.png');
end
nodes = readtable(fullfile(plotDirectory,'figure_2_5.csv'),'TextType','string');
beams = readtable(fullfile(plotDirectory,'chunk_beams.csv'));
shell = readtable(fullfile(plotDirectory,'chunk_shell_vertices.csv'));
original = nodes{:,{'original_x_m','original_y_m','original_z_m'}};
final = nodes{:,{'final_x_m','final_y_m','final_z_m'}};
skin = shell{:,{'x_m','y_m','z_m'}};
snapped = nodes.category == "snapped_to_shell_projection";
ordinary = nodes.category == "within_mpc_tolerance";
rejected = nodes.category == "rejected_snap_beam_length";
beyond = nodes.category == "beyond_snap_tolerance";
assert(height(nodes)==348 && height(beams)==399 && height(shell)==2139);
assert(nnz(snapped)==31 && nnz(ordinary)==50 && nnz(rejected)==21 && nnz(beyond)==246);
move = vecnorm(final-original,2,2);
assert(all(move(~snapped)==0) && all(move(snapped)>0));
assert(nnz(nodes.full_lattice_degree(snapped)==2)==29 && nnz(nodes.full_lattice_degree(snapped)==4)==2);
[foundA,indexA] = ismember(beams.grid_a,nodes.grid_id);
[foundB,indexB] = ismember(beams.grid_b,nodes.grid_id);
assert(all(foundA & foundB));
affected = snapped(indexA) | snapped(indexB);
assert(nnz(affected)==60);
assert(all(shell.local_vertex==repmat((1:3)',height(shell)/3,1)));
faces = reshape(1:height(shell),3,[])';
receivingFaces = faces(shell.receiving_for_visible_node(1:3:end)~=0,:);
assert(size(receivingFaces,1)==75);
% Rigid local frame: mean skin normal (+global z), global x projected onto
% that plane, and their right-handed transverse direction. No deformation.
normals = cross(skin(2:3:end,:)-skin(1:3:end,:),skin(3:3:end,:)-skin(1:3:end,:),2);
normals(normals(:,3)<0,:) = -normals(normals(:,3)<0,:);
normal = sum(normals,1); normal = normal/norm(normal);
horizontal = [1 0 0]-normal(1)*normal; horizontal = horizontal/norm(horizontal);
transverse = cross(normal,horizontal);
basis = [horizontal(:),transverse(:),normal(:)];
origin = mean(skin,1);
assert(norm(basis'*basis-eye(3),'fro')<1e-12);
original = (original-origin)*basis*1000;
final = (final-origin)*basis*1000;
skin = (skin-origin)*basis*1000;
extent = [min([original;final;skin],[],1);max([original;final;skin],[],1)];
padding = [2.5 2.5 2];
limits = extent+[-padding;padding];
% Slight in-plane rotation separates rows stacked in the old side view.
sideBasis = [cosd(5) 0;sind(5) 0;0 1];
originalSide = original*sideBasis; finalSide = final*sideBasis;
skinSide = skin*sideBasis;
sideExtent = [min([originalSide;finalSide;skinSide],[],1);max([originalSide;finalSide;skinSide],[],1)];
sideLimits = sideExtent+[-2.5 -2;2.5 2];
zoomLimits = [-24 -4;-8 .6];
visibleZoom = snapped & originalSide(:,1)>=zoomLimits(1,1) & originalSide(:,1)<=zoomLimits(2,1);
blue = [0.10 0.36 0.67]; orange = [0.85 0.32 0.04]; paleOrange = [.95 .74 .57]; ink = [0.19 0.20 0.22];
markerEdge = [0.38 0.17 0.04];
fig = figure('Visible','off','Color','w','Units','pixels','Position',[50 50 1600 1500]);
set(fig,'DefaultAxesFontName','Arial','DefaultTextFontName','Arial',...
    'DefaultTextInterpreter','none','DefaultLegendInterpreter','none','PaperPositionMode','auto');
annotation(fig,'textbox',[.045 .943 .91 .04],'String','Local lattice patch before and after mesh repair',...
    'EdgeColor','none','FontSize',20,'FontWeight','bold','Color',ink,'Margin',0);
annotation(fig,'textbox',[.045 .904 .91 .03],...
    'String',sprintf('399 beams: 60 affected, 339 unchanged | 31 nodes move %.3f–%.3f mm | no added beams',min(move(snapped))*1000,max(move(snapped))*1000),...
    'EdgeColor','none','FontSize',12.5,'Color',ink,'Margin',0);
positions = [.06 .61 .40 .28;.55 .61 .40 .28;.06 .345 .40 .21;.55 .345 .40 .21;.06 .163 .40 .125;.55 .163 .40 .125];
panelNames = {'(a) Before nodal repair — oblique','(b) After nodal repair — oblique',...
    '(c) Before nodal repair — side','(d) After nodal repair — side',...
    '(e) Before — enlarged skin interface','(f) After — enlarged skin interface'};
for k=1:6
    ax = axes(fig,'Position',positions(k,:)); hold(ax,'on');
    state = original; stateSide=originalSide; fill = 'w';
    if mod(k,2)==0, state=final; stateSide=finalSide; fill=orange; end
    if k<=2
      patch(ax,'Vertices',skin,'Faces',faces,'FaceColor',[.74 .76 .78],...
        'FaceAlpha',.12,'EdgeColor',[.69 .71 .73],'EdgeAlpha',.42,'LineWidth',.30);
      patch(ax,'Vertices',skin,'Faces',receivingFaces,'FaceColor',blue,...
        'FaceAlpha',.36,'EdgeColor',[.55 .57 .59],'EdgeAlpha',.65,'LineWidth',.35);
      for group=1:2
        selected=~affected;lineColor=paleOrange;lineWidth=.75;
        if group==2, selected=affected;lineColor=orange;lineWidth=1.4;end
        for dimension=1:3
          edgeCoordinates{dimension}=reshape([state(indexA(selected),dimension),state(indexB(selected),dimension),nan(nnz(selected),1)]',[],1); %#ok<SAGROW>
        end
        plot3(ax,edgeCoordinates{1},edgeCoordinates{2},edgeCoordinates{3},'-','Color',lineColor,'LineWidth',lineWidth);
      end
      plot3(ax,state(snapped,1),state(snapped,2),state(snapped,3),'o',...
        'LineStyle','none','MarkerSize',5.8,'MarkerEdgeColor',markerEdge,'MarkerFaceColor',fill,'LineWidth',1.25);
      xlim(ax,limits(:,1)'); ylim(ax,limits(:,2)'); zlim(ax,limits(:,3)');
      daspect(ax,[1 1 1]); camproj(ax,'orthographic');
      view(ax,[32 20]); xlabel(ax,'Local x (mm)'); ylabel(ax,'Local y (mm)'); zlabel(ax,'Local normal (mm)');
    else
      patch(ax,'Vertices',skinSide,'Faces',faces,'FaceColor',[.74 .76 .78],...
        'FaceAlpha',.12,'EdgeColor',[.69 .71 .73],'EdgeAlpha',.55,'LineWidth',.4);
      patch(ax,'Vertices',skinSide,'Faces',receivingFaces,'FaceColor',blue,...
        'FaceAlpha',.36,'EdgeColor',blue,'EdgeAlpha',.65,'LineWidth',.55);
      for group=1:2
        selected=~affected;lineColor=paleOrange;lineWidth=.8;
        if group==2, selected=affected;lineColor=orange;lineWidth=1.5;end
        for dimension=1:2
          edgeCoordinates{dimension}=reshape([stateSide(indexA(selected),dimension),stateSide(indexB(selected),dimension),nan(nnz(selected),1)]',[],1); %#ok<SAGROW>
        end
        plot(ax,edgeCoordinates{1},edgeCoordinates{2},'-','Color',lineColor,'LineWidth',lineWidth);
      end
      markerSize=5.8;if k>4,markerSize=7;end
      plot(ax,stateSide(snapped,1),stateSide(snapped,2),'o','LineStyle','none',...
        'MarkerSize',markerSize,'MarkerEdgeColor',markerEdge,'MarkerFaceColor',fill,'LineWidth',1.25);
      bounds=sideLimits;if k>4,bounds=zoomLimits;end
      xlim(ax,bounds(:,1)');ylim(ax,bounds(:,2)');daspect(ax,[1 1 1]);
      xlabel(ax,'In-plane position h (mm)');ylabel(ax,'Local normal (mm)');
      if k<=4
        rectangle(ax,'Position',[zoomLimits(1,:) diff(zoomLimits,1,1)],'EdgeColor',ink,'LineStyle','--','LineWidth',.8);
      end
    end
    set(ax,'FontSize',10,'Color','none','Box','off','TickDir','out',...
        'XColor',ink,'YColor',ink,'ZColor',ink,'GridColor',[.88 .88 .88]);
    grid(ax,'off');
    title(ax,panelNames{k},'FontSize',14,'FontWeight','bold','Units','normalized','Position',[.5 1.065 0]);
end
legendAx=axes(fig,'Position',[.05 .083 .90 .04],'Visible','off'); hold(legendAx,'on');
h1=plot(legendAx,nan,nan,'-','Color',orange,'LineWidth',1.8);
h2=plot(legendAx,nan,nan,'-','Color',[.64 .66 .68],'LineWidth',.7);
h3=plot(legendAx,nan,nan,'o','Color',markerEdge,'MarkerFaceColor','w','LineWidth',1.3);
h4=plot(legendAx,nan,nan,'o','Color',markerEdge,'MarkerFaceColor',orange,'LineWidth',1.3);
h5=plot(legendAx,nan,nan,'s','Color',[.55 .57 .59],'MarkerFaceColor',blue,'MarkerSize',8);
h6=plot(legendAx,nan,nan,'-','Color',paleOrange,'LineWidth',1.3);
legend(legendAx,[h1 h6 h2 h5 h3 h4],{'60 affected beams','339 unchanged beams','Skin-element edges',...
    'Receiving skin elements','Node before repair','Same node after repair'},...
    'Orientation','horizontal','NumColumns',3,'Location','north','Box','off','FontSize',11);
annotation(fig,'textbox',[.045 .014 .91 .052],'String',...
    {'Identical exported connectivity before and after repair; repositioning retained nodes is not load-induced deformation.',...
     'Overview panels retain all 399 beams. Dashed boxes locate the interface enlargements; lengths are unexaggerated.',...
     'Side direction: h = x cos(5°) + y sin(5°). Other nodes already contact the skin; there is no uniform clearance.',...
     'Data: accompanying node, beam and skin-element CSV files.'},...
    'EdgeColor','none','FontSize',10.5,'Color',ink,'Margin',0,'Interpreter','none');
drawnow;
% Preserve full canvas margins; exportgraphics tight-crops figure annotations.
[~,~,outputExtension] = fileparts(outputPath);
switch lower(outputExtension)
    case '.png'
        print(fig,outputPath,'-dpng','-r180');
    case '.pdf'
        print(fig,outputPath,'-dpdf','-bestfit');
    case '.svg'
        print(fig,outputPath,'-dsvg');
    otherwise
        close(fig);
        error('Output must have a .png, .pdf, or .svg extension.');
end
fprintf('Exported: %s\n',outputPath);
fprintf('Origin_m: %.15g %.15g %.15g\n',origin);
fprintf('Basis columns [horizontal transverse normal]:\n'); disp(basis);
fprintf('Displacement_mm range: %.12g %.12g\n',min(move(snapped))*1000,max(move(snapped))*1000);
fprintf('Affected beams: %d; unchanged beams: %d; moved nodes in enlargement: %d\n',nnz(affected),nnz(~affected),nnz(visibleZoom));
close(fig);
end

File: data/figures/chapter02/figure_2_5/plot_2_5.py
"""Reproduce Figure 2.5 using only the three CSV files beside this script.

Requirements: Python 3.10+, numpy, matplotlib. Run: python plot_2_5.py
The rigid frame, data selection, palette and six views match plot_2_5.m;
renderer-specific typography and 3-D depth ordering can differ.
"""
from pathlib import Path
import argparse
import csv
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from matplotlib.collections import LineCollection, PolyCollection
from matplotlib.lines import Line2D
from matplotlib.patches import Patch
from mpl_toolkits.mplot3d.art3d import Line3DCollection, Poly3DCollection

OUT = Path(__file__).resolve().parent

def read_csv(name):
    with (OUT/name).open(newline='',encoding='utf-8') as stream:
        return list(csv.DictReader(stream))

def main(output_path=None):
    nodes=read_csv('figure_2_5.csv'); beams=read_csv('chunk_beams.csv'); shell=read_csv('chunk_shell_vertices.csv')
    coords=lambda kind:np.array([[float(row[f'{kind}_{axis}_m']) for axis in 'xyz'] for row in nodes])
    original=coords('original'); final=coords('final')
    skin=np.array([[float(row[f'{axis}_m']) for axis in 'xyz'] for row in shell])
    categories=np.array([row['category'] for row in nodes]); snapped=categories=='snapped_to_shell_projection'
    assert (len(nodes),len(beams),len(shell))==(348,399,2139)
    assert tuple(np.sum(categories==kind) for kind in ['snapped_to_shell_projection','within_mpc_tolerance','rejected_snap_beam_length','beyond_snap_tolerance'])==(31,50,21,246)
    movement=np.linalg.norm(final-original,axis=1)
    assert np.all(movement[~snapped]==0)
    assert np.all(movement[snapped]>0)
    degrees=np.array([int(row['full_lattice_degree']) for row in nodes])
    assert ((degrees[snapped]==2).sum(),(degrees[snapped]==4).sum())==(29,2)
    lookup={int(row['grid_id']):i for i,row in enumerate(nodes)}
    endpoints=np.array([[lookup[int(row['grid_a'])],lookup[int(row['grid_b'])]] for row in beams])
    affected=snapped[endpoints].any(axis=1)
    assert affected.sum()==60
    assert all(int(row['local_vertex'])==i%3+1 for i,row in enumerate(shell))
    triangles=skin.reshape(-1,3,3)
    receiving=np.array([int(row['receiving_for_visible_node'])!=0 for row in shell[::3]])
    assert np.sum(receiving)==75
    normals=np.cross(triangles[:,1]-triangles[:,0],triangles[:,2]-triangles[:,0])
    normals[normals[:,2]<0]*=-1
    normal=normals.sum(axis=0);normal/=np.linalg.norm(normal)
    horizontal=np.array([1.,0.,0.])-normal[0]*normal;horizontal/=np.linalg.norm(horizontal)
    transverse=np.cross(normal,horizontal);basis=np.column_stack([horizontal,transverse,normal])
    origin=skin.mean(axis=0)
    assert np.linalg.norm(basis.T@basis-np.eye(3))<1e-12
    original=(original-origin)@basis*1000;final=(final-origin)@basis*1000;skin=(skin-origin)@basis*1000
    triangles=skin.reshape(-1,3,3)
    extent=np.vstack([original,final,skin]); limits=np.vstack([extent.min(axis=0)-[2.5,2.5,2],extent.max(axis=0)+[2.5,2.5,2]])
    # Rotate within the skin plane only. This separates depth-stacked lattice
    # rows without magnifying the movement or dropping any overview element.
    side_angle=np.deg2rad(5.)
    side_basis=np.column_stack([[np.cos(side_angle),np.sin(side_angle),0.],[0.,0.,1.]])
    original_side=original@side_basis; final_side=final@side_basis
    skin_side=skin@side_basis; triangles_side=skin_side.reshape(-1,3,2)
    side_extent=np.vstack([original_side,final_side,skin_side])
    side_limits=np.vstack([side_extent.min(0)-[2.5,2],side_extent.max(0)+[2.5,2]])
    zoom_limits=np.array([[-24.,-4.],[ -8.,.6]])
    visible_zoom=snapped & (original_side[:,0]>=zoom_limits[0,0]) & (original_side[:,0]<=zoom_limits[1,0])
    blue=(.10,.36,.67);orange=(.85,.32,.04);pale_orange=(.95,.74,.57);ink=(.19,.20,.22);outline=(.38,.17,.04)
    plt.rcParams.update({'font.family':'sans-serif','font.sans-serif':['Arial','DejaVu Sans'],'font.size':11,'text.color':ink,'axes.labelcolor':ink,'axes.edgecolor':(.55,.56,.57),'xtick.color':ink,'ytick.color':ink})
    fig=plt.figure(figsize=(16,15),facecolor='white')
    fig.text(.045,.972,'Local lattice patch before and after mesh repair',fontsize=20,weight='bold',va='top')
    fig.text(.045,.933,f'399 beams: 60 affected, 339 unchanged | 31 nodes move {movement[snapped].min()*1000:.3f}–{movement[snapped].max()*1000:.3f} mm | no added beams',fontsize=12.5,va='top')
    panel_positions=[(.045,.61,.43,.28),(.535,.61,.43,.28),(.065,.345,.39,.21),(.555,.345,.39,.21),(.065,.163,.39,.125),(.555,.163,.39,.125)]
    headings=['(a) Before nodal repair — oblique','(b) After nodal repair — oblique','(c) Before nodal repair — side','(d) After nodal repair — side','(e) Before — enlarged skin interface','(f) After — enlarged skin interface']
    for k,(position,heading) in enumerate(zip(panel_positions,headings)):
        state=original if k%2==0 else final
        marker_face='white' if k%2==0 else orange
        if k<2:
            ax=fig.add_axes(position,projection='3d',proj_type='ortho')
            ax.add_collection3d(Poly3DCollection(triangles,facecolors=(.74,.76,.78,.12),edgecolors=(.69,.71,.73,.42),linewidths=.3))
            ax.add_collection3d(Poly3DCollection(triangles[receiving],facecolors=(*blue,.36),edgecolors=(.55,.57,.59,.65),linewidths=.35))
            ax.add_collection3d(Line3DCollection(state[endpoints[~affected]],colors=pale_orange,linewidths=.75))
            ax.add_collection3d(Line3DCollection(state[endpoints[affected]],colors=orange,linewidths=1.4))
            ax.scatter(*state[snapped].T,s=30,facecolors=marker_face,edgecolors=outline,linewidths=1.25,depthshade=False)
            ax.set(xlim=limits[:,0],ylim=limits[:,1],zlim=limits[:,2],xlabel='Local x (mm)',ylabel='Local y (mm)',zlabel='Local normal (mm)')
            ax.set_box_aspect(limits[1]-limits[0],zoom=1.2);ax.view_init(elev=20,azim=-58)
            for axis in [ax.xaxis,ax.yaxis,ax.zaxis]:axis.pane.set_visible(False)
            ax.grid(False);ax.tick_params(labelsize=9,pad=0)
            ax.set_title(heading,fontsize=14,weight='bold',pad=7)
        else:
            ax=fig.add_axes(position)
            state_side=original_side if k%2==0 else final_side
            ax.add_collection(PolyCollection(triangles_side,facecolors=(.74,.76,.78,.12),edgecolors=(.69,.71,.73,.55),linewidths=.4))
            ax.add_collection(PolyCollection(triangles_side[receiving],facecolors=(*blue,.36),edgecolors=(*blue,.65),linewidths=.55))
            ax.add_collection(LineCollection(state_side[endpoints[~affected]],colors=pale_orange,linewidths=.8))
            ax.add_collection(LineCollection(state_side[endpoints[affected]],colors=orange,linewidths=1.5))
            ax.scatter(state_side[snapped,0],state_side[snapped,1],s=30 if k<4 else 45,facecolors=marker_face,edgecolors=outline,linewidths=1.25,zorder=4)
            bounds=side_limits if k<4 else zoom_limits
            ax.set(xlim=bounds[:,0],ylim=bounds[:,1],xlabel='In-plane position h (mm)',ylabel='Local normal (mm)')
            ax.set_aspect('equal',adjustable='box');ax.spines[['right','top']].set_visible(False)
            ax.set_title(heading,fontsize=14,weight='bold',pad=12)
            if k<4:
                ax.add_patch(plt.Rectangle(zoom_limits[0],*(zoom_limits[1]-zoom_limits[0]),fill=False,edgecolor=ink,linewidth=.8,linestyle='--',zorder=5))
    handles=[Line2D([],[],color=orange,lw=1.8,label='60 affected beams'),Line2D([],[],color=pale_orange,lw=1.3,label='339 unchanged beams'),Line2D([],[],color=(.64,.66,.68),lw=.7,label='Skin-element edges'),Patch(facecolor=blue,edgecolor=blue,label='Receiving skin elements'),Line2D([],[],marker='o',ls='none',markeredgecolor=outline,markerfacecolor='white',label='Node before repair'),Line2D([],[],marker='o',ls='none',markeredgecolor=outline,markerfacecolor=orange,label='Same node after repair')]
    fig.legend(handles=handles,loc='lower center',bbox_to_anchor=(.5,.083),ncol=3,frameon=False,fontsize=11)
    notes=['Identical exported connectivity before and after repair; repositioning retained nodes is not load-induced deformation.',
           'Overview panels retain all 399 beams. Dashed boxes locate the interface enlargements; lengths are unexaggerated.',
           'Side direction: h = x cos(5°) + y sin(5°). Other nodes already contact the skin; there is no uniform clearance.',
           'Data: accompanying node, beam and skin-element CSV files.']
    fig.text(.045,.065,'\n'.join(notes),fontsize=10.5,va='top',linespacing=1.4)
    output_path=Path(output_path) if output_path is not None else OUT/'figure_2_5_python.png'
    if output_path.suffix.lower() not in {'.png','.pdf','.svg'}:
        raise ValueError('Output must have a .png, .pdf, or .svg extension.')
    output_path.parent.mkdir(parents=True,exist_ok=True)
    fig.savefig(output_path,dpi=180,facecolor='white')
    plt.close(fig)
    print(f'Exported {output_path} from the three supplied CSV files.')
    print('Origin_m:',origin.tolist());print('Basis:',basis.tolist())
    print('Affected beams:',int(affected.sum()),'Unchanged beams:',int((~affected).sum()),'Moved nodes in enlargement:',int(visible_zoom.sum()))

if __name__=='__main__':
    arguments=argparse.ArgumentParser(description=__doc__)
    arguments.add_argument('--output',type=Path,help='Output PNG, PDF or SVG path; default: figure_2_5_python.png beside this script.')
    main(arguments.parse_args().output)
